File: 909_SnakesAndLadders.java
    /*  Explanation
    # Leetcode problem link : https://leetcode.com/problems/minesweeper/
    Time Complexity for operators : o(n*n) 
    Extra Space Complexity for operators : o(n*n) for move array
    Did this code successfully run on Leetcode : NA
    Any problem you faced while coding this : No
# Your code here along with comments explaining your approach
        # Basic approach : 
        # Optimized approach: 
                              
            # 1. 
                    A) 


    */ 

class Solution {
    public int snakesAndLadders(int[][] board) {
        
        if(board == null || board.length == 0 || board[0].length == 0)
            return 0;
        
        int n = board.length;
        int m = board[0].length;
        int move[] = new int[n*m];
        
        
        int even = 0;
        int index = 0;
        int i = n-1;
        int j = 0;
        
        while(i>=0 && j>=0){
            
            if(board[i][j] == -1){
                move[index] = -1;
            }else{
                move[index] = board[i][j] - 1;
            }
            
            // use even variable to move foawrd in the given 2 d matrix
            if(even % 2 == 0){
                j += 1;
            }else{
                j -= 1;
            }
            
            if(j >= m){
                even += 1;
                i -= 1;
                j -= 1;
            }else if (j < 0){
                even += 1;
                i -= 1;
                j += 1;
            }
            
            index += 1;
        }
        
        Queue<Integer> queue = new LinkedList<>();
        int minLevel = 0;
        queue.add(0);
        move[0] = -2;
        
        while(!queue.isEmpty()){
            int size = queue.size();
            
            for(int x = 0; x < size; x++){
                int front = queue.poll();
                
                if(front == n*m -1){
                    return minLevel;
                }
                
                for(int y = 1; y<=6;y++){
                    int nextMove = front+y;
                    
                    if(nextMove < (n*m) && move[nextMove] != -2){
                        if(move[nextMove] == -1){
                            queue.add(nextMove);
                        }else{
                            queue.add(move[nextMove]);
                        }
                        
                        move[nextMove] = -2;
                    }
                }
            }
            
            minLevel += 1;
        }
        
        return -1;
    }
}

File: 909-SnakesandLadders.cpp
/* 
    Time Complexity                              :  O(2*N*N) ~= O(N*N) 
    Space Complexity                             :  O(N*N) to store the values in the moves array
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std;  

class Solution {
public:
    int snakesAndLadders(vector<vector<int>>& board) {
        int n=board.size();
        vector<int> moves(n*n,-1);
        bool dir=true;
        int r=n-1,c=0,idx=0;
        while(r>=0 and c>=0) {
            if(board[r][c] == -1) {
                moves[idx] = -1;
            } else {
                moves[idx] = board[r][c] - 1;
            }
            idx++;
            
            if(dir) {
                c++;
                if(c == n) {
                    c--;
                    r--;
                    dir = false;
                }
            } else {
                c--;
                if(c == -1) {
                    c++;
                    r--;
                    dir = true;
                }
            }
            
        }
        
        // implement bfs
        queue<int> q;
        q.push(0);
        moves[0] = -2;
        int count=0;
        while(!q.empty()) {
            int sz = q.size();
            for(int i=0;i<sz;i++) {
                int id=q.front();q.pop();
                for(int j=1;j<=6;j++) {
                    int nid = id + j;
                    if(nid == n*n - 1 or moves[nid] == n*n - 1) 
                        return count+1;
                    if(nid >= n*n)
                        continue;
                    
                    if(moves[nid] != -2) {
                        if(moves[nid] == -1) {
                            q.push(nid);
                        } else {
                            q.push(moves[nid]);
                        }
                        moves[nid] = -2;
                    }
                }
            }
            count++;
        }
        
        return -1;
    }
};